llvm/include/llvm/CodeGen/SelectionDAG.h

```diff
@@ -1595,6 +1595,11 @@ class SelectionDAG {
   /// the target's desired shift amount type.
   SDValue getShiftAmountOperand(EVT LHSTy, SDValue Op);
 
+  /// Get a partial reduction SD node for the DAG. This is done when the input
+  /// and output types can be legalised for wide add(s) or dot product(s)
+  SDValue getPartialReduceAddSDNode(SDLoc DL, SDValue Chain, SDValue Acc,
+                                    SDValue Input);
+
   /// Create the DAG equivalent of vector_partial_reduce where Op1 and Op2 are
   /// its operands and ReducedTY is the intrinsic's return type.
   SDValue getPartialReduceAdd(SDLoc DL, EVT ReducedTy, SDValue Op1,
```

Comment:
I don't think there should be `DAG.getPartialReduceAddSDNode()` and `DAG.getPartialReduceAdd()`. 

I think `DAG.getPartialReduceAdd()` should always return the ISD node, and then in `SelectionDAGLegalize::ExpandNode(SDNode *Node)` in `LegalizeDAG.cpp`, the ISD node can be expanded (how it currently is in `DAG.getPartialReduceAdd()`) for targets that don't have a more specified lowering. 

Note: In `TargetLoweringBase::initActions()` you should probably set the default legalization action for `PARTIAL_REDUCE_ADD` to `Expand`. Then in  `AArch64ISelLowering` I think you'll want to set the action to `Custom` for supported types and handle the lowering to dsot/wide adds there.

Reply:
Changed the name of `getPartialReduceAdd` to `expandPartialReduceAdd`. For now, I'll try to carry on using the DAG combines, although I may need to take the approach described in your comment when legalisation is implemented.